docs/docs/reference/other-new-features/indentation.md

```diff
@@ -187,28 +187,102 @@ println(".")
+ - If the statement is a `val` definition binding a pattern, then `s` must be `val`.
+ - If the statement is a package clause that refers to package `p`, then `s` must be the same identifier `p`.
+ - If the statement is an `if`, `while`, `for`, `try`, or `match` statement, then `s` must be that same token.
+
+For instance, the following end markers are all legal:
+ ```scala
+  package p1.p2:
+
+    abstract class C():
+
+      def this(x: Int) =
+        this()
+        if x > 0 then
+          val a :: b =
+            x :: Nil
+          end val
+          var y =
+            x
+          end y
+          while y > 0 do
+            println(y)
+            y -= 1
+          end while
+          try
+            x match
+              case 0 => println("0")
+              case _ =>
+            end match
+          finally
+            println("done")
+          end try
+        end if
+      end this
+
+      def f: String
+    end C
+
+    object C:
+      given C =
+        new C:
+          def f = "!"
+          end f
+        end new
+      end given
+    end C
+
+    extension on (x: C):
+      def ff: String = x.f ++ x.f
+    end extension
+
+  end p2
 ```
-If `end` is followed by a reserved word, the compiler checks that the marker closes an indentation region belonging to a construct that starts with the reserved word. If it is followed by an identifier _id_, the compiler checks that the marker closes a definition
-that defines _id_ or a package clause that refers to _id_.
 
-`end` itself is a soft keyword. It is only treated as an `end` marker if it
-occurs at the start of a line and is followed by an identifier or one of the reserved words above.
+#### When to Use End Markers
+
+It is recommended that `end` markers are used for code where the extent of an indentation region is not immediately apparent "at a glance". People will have different preferences what this means, but one can nevertheless give some guidelines that stem from experience. An end marker makes sense if
+
+ - the construct contains blank lines, or
+ - the construct is long, say 15-20 lines or more,
+ - the construct ends heavily indented, say 4 indentation levels or more.
 
-It is recommended that `end` markers are used for code where the extent of an indentation region is not immediately apparent "at a glance". Typically this is the case if an indentation region spans 20 lines or more.
+If none of these criteria apply, it's often better to not use an end marker since the code will be just as clear and more concise. If there are several ending regions that satisfy one of the criteria above, we usually need an end marker only for the outermost closed reason. So cascades of end markers as in the example above are usually better avoided.
+
+#### Syntax
+
+```
+EndMarker         ::=  ‘end’ EndMarkerTag    -- when followed by EOL
+EndMarkerTag      ::=  id | ‘if’ | ‘while’ | ‘for’ | ‘match’ | ‘try’
```

Comment:
The `try` seems to be a display problem, it looks fine when we render the page separately.